### DISCUSSION TITLE
SQLalchemy with HA Postgres

### DISCUSSION THREAD

**Discussion by caniko** [fastapi-users/fastapi-users#977] [Q&A] [answered]
I have a postgres cluster with HA enabled. I would like to direct `UserCreate` and `UserUpdate` to the master, and `UserRead` to the replicas. I didn't have a problem with this before `10.0` because [`tortoise-orm`](https://tortoise-orm.readthedocs.io/en/latest/router.html?highlight=master#config-router) had support that played really nicely with `fastapi-users`, this is not the case with `SQLalchemy`, where I have two engines like mentioned [here](https://github.com/sqlalchemy/sqlalchemy/issues/5815#issuecomment-1119104934). However, I can only pass one engine to the `UserManager`.

**Comment by frankie567** [comment]
Well, that's quite a complex thing to handle. I see two solutions for you:

1. The easiest way: adapt the Tortoise DB adapter to the new version. You can base on [the previous one](https://github.com/fastapi-users/fastapi-users-db-tortoise).
    * It should be quite straighforward. You can compare for example the changes for SQLAlchemy : https://github.com/fastapi-users/fastapi-users-db-sqlalchemy/compare/v3.0.1..main
    * The most notable change is the addition of two new methods `add_oauth_account` and `update_oauth_account`.
2. A more complex way would be to overload the `SQLAlchemyUserDatabase` so it accepts two instances of AsyncSession: a "read" one and a "write" one. Then, you would need to overload each method so that it uses the right session depending on the context.

**Reply by caniko** [reply]
I can, but why not add a feature where method nr. 2 is supported natively by `fastapi-user`? It really isn't that complex, IMHO: We would add a `reader` session, and only change the `read` method.

I don't want to write too much code dependent on `fastapi-user` considering how often things change around here. And shouldn't there be support for HA? I can make a PR if you are interested.

**Reply by frankie567** [reply]
> I can, but why not add a feature where method nr. 2 is supported natively by fastapi-user? It really isn't that complex, IMHO

In 3+ years of existence of the library, it's the first time I have a request regarding HA databases with write/read-only nodes. So yeah, that may not be complex to implement but I'm not really willing to maintain a feature that even me won't use.

Besides, the configuration/installation is already quite complicated, adding yet another layer of complexity would be fearsome for everyone.

> I don't want to write too much code dependent on fastapi-user considering how often things change around here.

Be assured that I'm doing my *best*. I'm 100% alone on this project so I can't commit to support every specific use-case.

What I'm trying to do is to design FastAPI Users so it can be highly customizable and let developers implement their own database adapters, their own authentication strategies, their own routers... Even frameworks like Django don't offer this level of customization.

If you have very specific requirements, maybe an out-of-the-shelf solution is not appropriate. FastAPI has a very good tutorial on [how to implement authentication](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/), and you can even copy-paste parts of FastAPI Users, that's 100% valid.

**Reply by caniko** [reply]
I understand that you have a very specfic view on what `fastapi-user` should be, but I ask you that you widen your scope to garner more contributors and maintainers for the project. Long-term health, right?

> Be assured that I'm doing my _best_. I'm 100% alone on this project so I can't commit to support every specific use-case.

I just voiced my willingness to help, and you are telling me you are alone. [`HA`](https://postgres-operator.readthedocs.io/en/latest/) is a must in enterprise/high traffic applications.

> What I'm trying to do is to design FastAPI Users so it can be highly customizable and let developers implement their own database adapters, their own authentication strategies, their own routers...

The beauty of Python is that this approach is no longer nescessary, you can simply add support for everything. Yes, a line in the sand must be drawn at some point, but drawing it at `HA` to make a point is outlandish to me.

> Even frameworks like Django don't offer this level of customization.

Django is very accomadating towards every database system architecture as long as it is supported by the underlying driver, same should be achieved here, right? Your method nr. 2 compared to Django is far more complex, just look [here](https://medium.com/@omi10859/master-slave-db-in-django-with-psql-61dc10bff4c4) and [here](https://docs.djangoproject.com/en/4.0/topics/db/multi-db/#using-routers); almost entirely in yaml format. Beautiful. I am not asking for exactly this, but the `fastapi-user` spin on the implementation.

> In 3+ years of existence of the library, it's the first time I have a request regarding HA databases with write/read-only nodes.

There is a first time for everything. Please be more accomodating to requests that go beyond your own use-case.

**Comment by caniko** [comment, marked as answer]
I really don't understand why this can't be added to the package. I made my own implementation:
```python
from typing import Optional

from fastapi_users.models import UP
from fastapi_users_db_sqlalchemy import SQLAlchemyUserDatabase
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.sql import Select


class HowFossShouldBeButIsntSQLAlchemyUserDatabase(SQLAlchemyUserDatabase):
    read_session: AsyncSession

    def __init__(self, read_session: AsyncSession, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.read_session = read_session

    async def _get_user(self, statement: Select) -> Optional[UP]:
        results = await self.read_session.execute(statement)
        user = results.first()
        if user is None:
            return None

        return user[0]
```

**Reply by frankie567** [reply]
@caniko, I don't want to be rude but your messages are really not kind and respectful towards the work I'm doing and the time I spend to answer to everyone here.

I'm glad you figured out the implementation. It's short and clean. I don't know why you're so upset to just keep it in your project and make it evolve for your own needs.

**Reply by caniko** [reply]
I think you mean to say you don't like what I wrote. To refer to me as unkind and disrespectful when I provide you with constructive feedback and argue in favor of something you don't agree with is ignorant, and consequently rude; so you have failed in your mission I am afraid.

Yes, I am a little bitter from your lack of understanding, which I find warrented from the outcome. First you drop support for `tortoise-orm`, and then you drop support for database architectures. Yes, it is involuntary and implicit, but lacks thought and care for the community. I even suggested helping.

**Reply by frankie567** [reply]
True, I don't like what you wrote. I'm also concerned that on your side you don't make any effort to understand my position.

Ask *every* open-source maintainers out there: they can't accept all suggestions made by everyone. Assuming they do, they would end-up with a pile of features that are barely used by one or two people; with an unmaintainable codebase.

So yeah, I'm perfectly comfortable with declining suggestions I don't like or don't want to maintain. Even more considering that I provided you guidelines to implement it on your own.

True, I dropped support for Tortoise ORM. It was a burden to me. I don't really understand what you mean by dropping "database architecture support". Supporting your use-case was a happy consequence of Tortoise, but never was supported or mentioned anywhere.

Now for the last part: I do care for the community. Once again, I spend hours on Issues and Discussions answering the questions of everyone. I do it because I like to help and share my experience.

However, I don't owe anything to anyone. I'm providing hundreds hours worth of work for free. So yes, I'm keeping the right to drop things and decline ideas. I understand it can be frustrating, but that's how I can make open-source sustainable for me. 

If that's not your vision, you can fork the library and make your own version, that's allowed. Possibly, you'll be better than me in this role and be able to better accommodate the demands of your community.

**Reply by caniko** [reply]
I understand @frankie567, this is your safe space not ~mine~ ours. I was under the wrong impression, my apologies, happy developing [alone](https://github.com/fastapi-users/fastapi-users/discussions/977#discussioncomment-2700183).
